### ISSUE TITLE
AWS termination of node -- kick off new builds for affected jobs

### ISSUE THREAD

**Issue by bertjwregeer** [jenkinsci/ec2-fleet-plugin#45]
When AWS kills a node because of some external event (such as it wants that instance type back), then jobs running on those nodes will wait until timeout, or forever, while the node is marked offline.

Instead it would be nice that if a node disappears, for any reason, and it has currently active jobs on it, that 1. those jobs are killed and 2. that those jobs are "kicked".

In my case it's a bunch of Jenkins pipeline jobs, and it would be great to kick them so that another executor can hopefully pick up the job and complete it before the next node is kicked over by Amazon.

**Comment by terma** [comment]
We can handle EC2 Spot termination notice to stop executors upfront https://aws.amazon.com/blogs/aws/new-ec2-spot-instance-termination-notices/

**Comment by terma** [comment]
Related issue #87

**Comment by terma** [comment]
Way to execute code on Slave

https://wiki.jenkins.io/display/JENKINS/Making+your+plugin+behave+in+distributed+Jenkins
https://javadoc.jenkins.io/component/remoting/hudson/remoting/Channel.html#call-hudson.remoting.Callable-

**Comment by terma** [comment]
## Possible solution

### Idea

- From describe, fleet instances consume additional parameter ```status code``` which represents if an instance was marked for termination
- In case if any task in progress right now for this node, try to send them back to the queue

### Implementation

1. ```EC2Api.describeTerminated``` to check ```status code``` additionally to existent ```state``` in case of ```marked-for-termination``` or ```marked-for-stop``` assume node as terminated, so cloud logic will remove them from Jenkins
1. Upgrade ```EC2FleetCloud.removeNode``` to disable accepting for new jobs
  - Looks like we don't need to explicitly 
1. Upgrade ```EC2FleetCloud.removeNode``` to check if any job in progress and send it back to queue (this item requires additional research)

### Links

- https://aws.amazon.com/blogs/aws/new-ec2-spot-instance-termination-notices/
- https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-bid-status.html
- https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeSpotInstanceRequests.html

**Comment by bertjwregeer** [comment]
> Upgrade EC2FleetCloud.removeNode to check if any job in progress and send it back to queue (this item requires additional research)

For this, just kicking off a new build of the affected project would work for us. Starting all of the stages from scratch, being able restart the stage we are currently on would be better, but the first step of just restarting it from scratch would be amazing.

**Comment by terma** [comment]
For now Jenkins doesn't provide automatic way to re-run job, as discussed https://groups.google.com/forum/?nomobile=true#!topic/jenkinsci-dev/7wVl1BTxiXs

**Comment by terma** [comment]
hi, I think #104 could help you, it allows EC2 Fleet Plugin auto re-submit failed (because of Spot interruption) jobs. As soon as it will be merged will update with plugin version. Thx.

**Comment by terma** [comment]
auto resubmit of failed jobs (EC2 instance termination) PR #104 released under [1.7.0](https://github.com/jenkinsci/ec2-fleet-plugin/releases/tag/ec2-fleet-1.7.0)

This change is small but really tricky, as there is no single way to do that. I tested it with my Jenkins on a few types of jobs but Jenkins is big and there are so many different types of Jobs and their configuration, so your feedback is welcome! Thx.

**Comment by bertjwregeer** [comment]
I've installed the update on our Jenkins server, I'll let you know how things fare over the next week or so.

**Comment by bertjwregeer** [comment]
Is there anything special I need to do to enable this feature? I've got a fleet node that is being removed (and is offline), but I am not seeing the job being kicked off.

Are there some logs I can log for? Something I can grep for?

**Comment by bertjwregeer** [comment]
> Cannot contact i-0c47129b827a6a8c2: hudson.remoting.ChannelClosedException: Channel "unknown": Remote call on i-0c47129b827a6a8c2 failed. The channel is closing down or has closed down
> Cannot contact i-0c47129b827a6a8c2: hudson.remoting.ChannelClosedException: Channel "unknown": Remote call on i-0c47129b827a6a8c2 failed. The channel is closing down or has closed down

Is the last two messages I see in one of the jobs (parallel jobs).

I am not seeing output from ec2-fleet plugin that is attempting to retry them.

**Comment by terma** [comment]
hm, it doesn't require any enabling, can I ask about Jenkins version?

**Comment by terma** [comment]
in log you should see smt like 
```java
LOGGER.info(" job " + task.getName() + " on " + getDisplayName() + " execution aborted because of slave EC2 instance was terminated, resubmitting");
```

**Comment by bertjwregeer** [comment]
Jenkins ver. 2.176.1 (latest `jenkinsci/blueocean` docker container)

I am not seeing "execution aborted" in any of my Jenkins logs.

**Comment by terma** [comment]
ok, let me test under your version.

**Comment by bertjwregeer** [comment]
> INFO: I/O error in channel i-0c47129b827a6a8c2
> INFO: fleet = Spot Fleet label = docker Fleet (docker) no longer has the instance i-0c47129b827a6a8c2, removing from Jenkins.

**Comment by terma** [comment]
Ok, I think I found a case, [Pipeline Durability](https://issues.jenkins-ci.org/browse/JENKINS-33761?page=com.atlassian.jira.plugin.system.issuetabpanels%3Aall-tabpanel), which prevents fail job, so plugin can detect disconnection, but continue and wait when node will be up, 

here is my pipeline run:
```
[Pipeline] {
[Pipeline] stage
[Pipeline] { (Stage 1)
[Pipeline] parallel
[Pipeline] { (Branch: a)
[Pipeline] { (Branch: b)
[Pipeline] stage
[Pipeline] { (a)
[Pipeline] stage
[Pipeline] { (b)
[Pipeline] sh
[Pipeline] sh
+ sleep 90
+ sleep 90
Cannot contact i-0d4ad3dd0c34b70b6: hudson.remoting.ChannelClosedException: Channel "unknown": Remote call on i-0d4ad3dd0c34b70b6 failed. The channel is closing down or has closed down
Cannot contact i-0d4ad3dd0c34b70b6: hudson.remoting.ChannelClosedException: Channel "unknown": Remote call on i-0d4ad3dd0c34b70b6 failed. The channel is closing down or has closed down
```

To confirm, do you see similar in your case?

**Comment by bertjwregeer** [comment]
Yes, I can confirm that this is during a pipeline and we have pipeline durability.

**Comment by terma** [comment]
ok, the second try, I hope last one, this kind of Jenkins is tricky, anyway **big** for testing report,

I added additional logic for Pipeline based jobs, as they don't fail in case of disconnect and wait for reconnecting which is not expected in Spot case (usually). Details in #107 

will publish fixed version later.

**Comment by terma** [comment]
Fix for **pipeline jobs** released under [1.7.1](https://github.com/jenkinsci/ec2-fleet-plugin/releases/tag/ec2-fleet-1.7.1)

**Comment by bertjwregeer** [comment]
\o/ this has now been deployed to our Jenkins instance. I'll keep you updated on how things go.

**Comment by romang12** [comment]
Hi guys,
My question concerns to usage of this feature for multi-configuration job.
For example, I have build that runs on different configurations (instances), and one of instance went offline during the build. EC2 Fleet plugin will recognize such issue and will re-run for this job.
Is it possible to perform retry only for failed configurations (not for all)?

**Comment by terma** [comment]
@romang12, yes, tgat's how plugin works, it dies resubmit only for failed nodes.

**Comment by romang12** [comment]
@terma This is configuration of my Fleet:

![image](https://user-images.githubusercontent.com/19771895/60454389-5fc18680-9c3c-11e9-95bf-cf7ca6d7d131.png)

Configuration failed due to AWS issue:
18:25:46 Slave went offline during the build
18:25:46 ERROR: Connection was broken: java.io.IOException: Unexpected termination of the channel
18:25:46 	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:50)
18:25:46 Caused by: java.io.EOFException
18:25:46 	at java.io.ObjectInputStream$PeekInputStream.readFully(ObjectInputStream.java:2335)
18:25:46 	at java.io.ObjectInputStream$BlockDataInputStream.readShort(ObjectInputStream.java:2804)
18:25:46 	at java.io.ObjectInputStream.readStreamHeader(ObjectInputStream.java:802)
18:25:46 	at java.io.ObjectInputStream.<init>(ObjectInputStream.java:299)
18:25:46 	at hudson.remoting.ObjectInputStreamEx.<init>(ObjectInputStreamEx.java:40)
18:25:46 	at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.read(AbstractSynchronousByteArrayCommandTransport.java:34)
18:25:46 	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:48)

But resubmit didn't work... Any idea?

**Comment by terma** [comment]
@romang12 please share version of jenkins, plugin and type of job

do you see in Jenkins logs smt from:
```
LOGGER.log(LOG_LEVEL, "Unexpected " + computer.getDisplayName() + " termination,  resubmit");
...
LOGGER.log(LOG_LEVEL, "Unexpected " + computer.getDisplayName() + " termination, resubmit " + task + " with actions " + actions);
...
LOGGER.log(LOG_LEVEL, "Unexpected " + computer.getDisplayName() + " termination, resubmit finished");
...
LOGGER.log(LOG_LEVEL, "Unexpected " + computer.getDisplayName() + " termination but resubmit disabled, no actions");
```

**Comment by romang12** [comment]
@terma 
Jenkins version: 1.609.3
EC2 Fleet Plugin 1.7.1
Job of type: Multi-configuration project

Jenkins log:

INFO: Unexpected Staging_Fleet_Runner_Private i-0c5f576255bdcd5e9 termination but resubmit disabled, no actions
INFO: Unexpected removing fleet node termination but resubmit disabled, no actions

**Comment by terma** [comment]
@romang12 

very interesting, 

Auto-resubmit could be disable by a few reasons: 
- disabled by configuration (looks like no)
- node is not offline (looks like online)
- termination reason is not ```OfflineCause.ChannelTermination```

Cannot reproduce this on my machine. Please use [1.7.2](https://github.com/jenkinsci/ec2-fleet-plugin/releases/tag/ec2-fleet-1.7.2) with additional log information, as soon as it fail again, please report log.

Btw, plugin supports Jenkins versions start from ```1.625.3```, how did you install it =) any way it doesn't seem like a problem for this particular case.

Thx,
Artem

**Comment by romang12** [comment]
@terma 
The requested URL /plugins/ec2-fleet/1.7.2/ec2-fleet.hpi was not found on this server.
:)

**Comment by terma** [comment]
@romang12 sorry my bad, I release it as regular version, you should be able to update it from Jenkins

**Comment by romang12** [comment]
@terma Thank you for your help!
I've upgraded to 1.7.2. Below please see the log concerning instance issue:

From my job log:

09:04:43 Slave went offline during the build
09:04:43 ERROR: Connection was broken: java.io.IOException: Unexpected termination of the channel
09:04:43 	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:50)

From Jenkins log:

SEVERE: I/O error in channel i-0504b38f435ba02cc
java.io.IOException: Unexpected termination of the channel
        at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:50)
		
INFO: Unexpected Staging_Fleet_Runner_Private i-0504b38f435ba02cc termination,  resubmit
Jul 02, 2019 9:04:43 AM hudson.model.Run execute

INFO: Check-My-Sanity-roman/Platform=linux,RELEASE=0,Role=private-runner,configXML=1 #14 main build action completed: FAILURE
Jul 02, 2019 9:04:43 AM com.amazon.jenkins.ec2fleet.EC2FleetAutoResubmitComputerLauncher afterDisconnect
INFO: Unexpected Staging_Fleet_Runner_Private i-0504b38f435ba02cc termination, resubmit hudson.matrix.MatrixConfiguration@42ea020f[Check-My-Sanity-roman/Platform=linux,RELEASE=0,Role=private-runner,configXML=1] with actions [hudson.matrix.MatrixChildParametersAction@71232b, hudson.matrix.MatrixConfiguration$ParentBuildAction@465de78, hudson.model.CauseAction@d04cda4, hudson.plugins.s3.S3CopyArtifact$EnvAction@3d813fa3, org.jenkinsci.plugins.envinject.EnvInjectPluginAction@7fd0476, hudson.plugins.git.util.BuildData@3accf21b[scmName=<null>,remoteUrls=[git@lab.vibelab.net:backend/IM2.git],buildsByBranchName={detached=Build #14 of Revision 87b4983f4297211940f88266359f41283807daf3 (detached)},lastBuild=Build #14 of Revision 87b4983f4297211940f88266359f41283807daf3 (detached)], hudson.plugins.git.GitTagAction@504d4133, hudson.scm.SCMRevisionState$None@2cb15e96, hudson.tasks.Fingerprinter$FingerprintAction@32ae5dad, com.sonyericsson.rebuild.RebuildAction@55700400]
Jul 02, 2019 9:04:43 AM com.amazon.jenkins.ec2fleet.EC2FleetAutoResubmitComputerLauncher afterDisconnect

INFO: Unexpected Staging_Fleet_Runner_Private i-0504b38f435ba02cc termination, resubmit finished
Jul 02, 2019 9:04:43 AM com.amazon.jenkins.ec2fleet.EC2FleetAutoResubmitComputerLauncher afterDisconnect

INFO: Reconnecting to instance: Staging_Fleet_Runner_Private i-0504b38f435ba02cc
Jul 02, 2019 9:04:44 AM hudson.slaves.SlaveComputer tryReconnect
INFO: Attempting to reconnect i-0504b38f435ba02cc
Jul 02, 2019 9:04:44 AM hudson.triggers.SafeTimerTask run
SEVERE: Timer task com.amazon.jenkins.ec2fleet.CloudNanny@571cbb02 failed
java.util.ConcurrentModificationException
        at java.util.HashMap$HashIterator.nextNode(HashMap.java:1429)

**Comment by terma** [comment]
@romang12 

based on provided log, I see that plugin did resubmit for your job:

```
INFO: Unexpected Staging_Fleet_Runner_Private i-0504b38f435ba02cc termination, resubmit hudson.matrix.MatrixConfiguration@42ea020f[Check-My-Sanity-roman/Platform=linux,RELEASE=0,Role=private-runner,configXML=1] with actions [hudson.matrix.MatrixChildParametersAction@71232b, hudson.matrix.MatrixConfiguration$ParentBuildAction@465de78, hudson.model.CauseAction@d04cda4, hudson.plugins.s3.S3CopyArtifact$EnvAction@3d813fa3, org.jenkinsci.plugins.envinject.EnvInjectPluginAction@7fd0476, hudson.plugins.git.util.BuildData@3accf21b[scmName=,remoteUrls=[git@lab.vibelab.net:backend/IM2.git],buildsByBranchName={detached=Build #14 of Revision 87b4983f4297211940f88266359f41283807daf3 (detached)},lastBuild=Build #14 of Revision 87b4983f4297211940f88266359f41283807daf3 (detached)], hudson.plugins.git.GitTagAction@504d4133, hudson.scm.SCMRevisionState$None@2cb15e96, hudson.tasks.Fingerprinter$FingerprintAction@32ae5dad, com.sonyericsson.rebuild.RebuildAction@55700400]
Jul 02, 2019 9:04:43 AM com.amazon.jenkins.ec2fleet.EC2FleetAutoResubmitComputerLauncher afterDisconnect

INFO: Unexpected Staging_Fleet_Runner_Private i-0504b38f435ba02cc termination, resubmit finished
Jul 02, 2019 9:04:43 AM com.amazon.jenkins.ec2fleet.EC2FleetAutoResubmitComputerLauncher afterDisconnect
```

second line of log related to [code](https://github.com/jenkinsci/ec2-fleet-plugin/blob/master/src/main/java/com/amazon/jenkins/ec2fleet/EC2FleetAutoResubmitComputerLauncher.java#L102) which does actual resubmit.

Did you see resubmitted job run? original one should be aborted and red, but new one should be submitted. 

Btw. I noticed during my local test, that autoresubmit for Multi Configuration type of jobs, doesn't do that gracefully, all actual jobs under Multi Configuration will be correctly resubmitted and reruned but main one which is controller, will be red at the end as it expecteds for example 1 jobs, but in fact 1 failed, and 1 resubmitted green, so it fails. Probably you see this problem?

Thx,

**Comment by romang12** [comment]
@terma 
Hi Artem,
Thank you very much for your effort!

I use Dynamic Axes for Multi-Configuration

![image](https://user-images.githubusercontent.com/19771895/60572909-3f4e1500-9d7f-11e9-96e0-a3dce19ad6ab.png)

Unfortunately, no controller (parent job) and not configuration (child) were not resubmitted in my case.

Currently , as a workaround, I use Naginator plugin, that permits me to automatically re-run failed jobs.

**Comment by terma** [comment]
@romang12 thx for details, I will try to simulate your env and test plugin, meantime I will update plugin documentation to recommend the use of Naginator as a workaround for the plugin.

**Comment by romang12** [comment]
@terma Hi Artem
I've installed the latest 1.9.0 version. It works well.
I noticed that labels stopped to appear on EC Fleet status

![image](https://user-images.githubusercontent.com/19771895/60830151-3d40e780-a1bf-11e9-88ff-143951006a02.png)

in spite of the fact that it does appear on Fleet configuration:

![image](https://user-images.githubusercontent.com/19771895/60830269-82651980-a1bf-11e9-9daf-c671f01795b0.png)

It worked for me in my previous version 1.1.9.

Do I need to reconfigure Fleet properties?

**Comment by terma** [comment]
Hm, interesting, I don't think you need to reconfigure it, as soon as you have labels in configuration they should be in widget, too. Let me check.

**Comment by terma** [comment]
@romang12 I can confirm bug, created issue #117 fix in progress

**Comment by romang12** [comment]
@terma 
It's possible, instead of labels, just to add Fleet name over Fleet ID. It will be pretty good :-)

**Comment by terma** [comment]
@romang12 

1. fix released under [1.9.1](https://github.com/jenkinsci/ec2-fleet-plugin/releases/tag/ec2-fleet-1.9.1)
1. Fleet name is possible, here draft issue for that #118 please take a look add/update details as you see this feature

**Comment by terma** [comment]
Looks like no more problem with this feature, thx @bertjwregeer for issue and @romang12 for testing. Going to close issue, feel free to reopen or comment if you have some problems or questions. 

**Note** In general case I suggest to create new issue instead of comment, so we can easly discuss and track them. thx.
